src/components/coffeechat.ts

```diff
@@ -0,0 +1,398 @@
+ * Since the ID's are normalized, we can return a 2D array of the tallies
+ */
+const loadMatched = async (notMatched: Map<string, number>): Promise<number[][]> => {
+  const db = await openDB();
+  const matched: number[][] = Array.from(Array(notMatched.size), () => new Array(4).fill(0));
+  const matches = (await db.all(`SELECT * FROM coffee_historic_matches`)) as historic_match[];
+  for (const { first_user_id, second_user_id } of matches) {
+    if (notMatched.has(first_user_id) && notMatched.has(second_user_id)) {
+      matched[notMatched.get(first_user_id)!][notMatched.get(second_user_id)!]++;
+      matched[notMatched.get(second_user_id)!][notMatched.get(first_user_id)!]++;
+    }
+  }
+  return matched;
+};
+
+/*
+ * Writes a new round of matches into HISTORY
+ * Use writeFutureMatches to save potential matches into future
+ */
+export const writeHistoricMatches = async (newMatches: string[][]): Promise<void> => {
+  const db = await openDB();
+
+  //insert (?,?,CURRENT_TIMESTAMP) into command based on number of rows, then populate params into second arg
+  //uses sqlite's builtin CURRENT_TIMESTAMP to get the current standard time
+  await db.run(
+    `INSERT INTO coffee_historic_matches (first_user_id, second_user_id, match_date) VALUES ${_.join(
+      newMatches.map(() => `(?,?, CURRENT_TIMESTAMP)`),
+      ','
+    )};`,
+    _.flatten(newMatches)
+  );
+};
+
+/*
+ * Wipes a database
+ */
+const wipeHistory = async (dbName: string): Promise<void> => {
+  const db = await openDB();
+  await db.run(`DELETE FROM ${dbName}`);
+};
+
+/*
+ * matching algorithm leveraging stable marriage
+ * takes in the users and their historical match data
+ * Returns a round of match results based on this algorithm:
+ * 1. Separate users randomly into 2 even sections, removing 1 user if list size is odd
+ * 2. build a preference list for each user, sorted by how little they have chatted with each user
+ * in the other section (fewest chatted user comes first, then second, etc)
+ * 3. Run the stable marriage library with the 2 sections and their preferences
+ * 4. Test this output to a finalOutput; If contender has a lower MaxDupe than the final, overwrite final
+ * 5. Run 1-4 RANDOM_ITERATIONS times; every time an overwrite occurs, restart the iteration count
+ * Return finalOutput
+ */
+const stableMatch = (userList: Map<string, number>, matched: number[][]): string[][] => {
+  let finalOutput: string[][] | undefined = undefined;
+  for (let i = 0; i < RANDOM_ITERATIONS; i++) {
+    //shuffle users, then separate into 2 sections
+    const notMatched = _.shuffle(Array.from(userList).map((name) => name[0]));
+    const A = notMatched.slice(0, Math.floor(notMatched.length / 2)).map((name) => new Person(name));
+    const B = notMatched.slice(Math.floor(notMatched.length / 2)).map((name) => new Person(name));
+
+    //attach preference list for each user by creating comparator which prioritizes user with less chats than the subject
```

Comment:
If I'm understanding correctly, maybe `with less chats than the subject` should be reworded to `that has less chats with the subject`?